### PR TITLE
Reorganize interface parsing to be more consistent with other declarations.

### PR DESCRIPTION
**Pull request by jonmeow**
The comments in parse_node_kind.def capture the change being made here.

Before:

```
//   _external_: DeclaredName
//     InterfaceBodyStart
//     _external_: statements
//   InterfaceBodyEnd
// InterfaceDefinition
```

After:

```
//     InterfaceIntroducer
//     DeclaredName
//   InterfaceDefinitionStart
//   _external_: declarations
// InterfaceDefinition
```

Really I just want to treat introduced things consistently. `var` defines my philosophy here: it doesn't always have a `DeclaredName`, so the `VarIntroducer` _must_ be the bounding node. By being consistent with that, I believe that overall the structure becomes easier to understand (that is, there are fewer inconsistencies to understand).

This also adds InterfaceDeclaration, since I think it can be predicted we'll have that, and it's helpful for making recover consistent with HandleDeclarationError.

Similarly, I'm also trying to standardize the loop processing a little with HandleDeclarationLoop. In the current approach, InterfaceDefinitionFinish isn't a necessary state, so I'm removing it.

### CODE REVIEW

File: toolchain/parser/parse_node_kind.def [hunk 1]
```diff
@@ -282,14 +282,18 @@ CARBON_PARSE_NODE_KIND_BRACKET(StructTypeLiteral,
...
+//     DeclaredName
+//   InterfaceDefinitionStart
+//   _external_: declarations
 // InterfaceDefinition
-CARBON_PARSE_NODE_KIND_CHILD_COUNT(InterfaceBodyStart, 0)
-CARBON_PARSE_NODE_KIND_BRACKET(InterfaceBody, InterfaceBodyStart)
-CARBON_PARSE_NODE_KIND_BRACKET(InterfaceDefinition, DeclaredName)
+//
+// The above is a definition; for a declaration, InterfaceDeclaration will end
+// it InterfaceDefinitionStart is for a definition.
```
Comment (zygoloid) [inline toolchain/parser/parse_node_kind.def]: I can't parse this comment.

File: toolchain/parser/parser.h [hunk 2]
```diff
@@ -262,6 +262,17 @@ class Parser {
   // parses should only need to look down a couple steps.
   auto GetDeclarationContext() -> DeclarationContext;
 
+  // Handles of recovery in a declaration, particularly before any possible
```
Comment (zygoloid) [inline toolchain/parser/parser.h]: Is there a missing or extra word near the start of this sentence?

**Review by zygoloid — APPROVED** [review]

File: toolchain/parser/parse_node_kind.def (hunk 1 above)
Comment (jonmeow) [reply toolchain/parser/parse_node_kind.def]: Trying to rephrase to help, along with the matching function comment. New text:

```
// The above is the structure for a definition; for a declaration,
// FunctionDefinitionStart and later nodes are removed and replaced by
// FunctionDeclaration.
```

File: toolchain/parser/parser.h (hunk 2 above)
Comment (jonmeow) [reply toolchain/parser/parser.h]: of -> error

**Pull request merged**